File: pbhhg_py/abstract_syntax.py
from __future__ import annotations

import abc
import dataclasses
import math
import typing
from typing import NamedTuple
import unicodedata


def _get_width(s: str) -> int:
    return sum(2 if unicodedata.east_asian_width(c) in "FW" else 1 for c in s)


class Metadata(NamedTuple):
    filename: str
    line_no: int
    start_col: int
    end_col: int
    line: str  # whole line

    def __str__(self) -> str:
        return (
            self.line
            + "\n"
            + " " * _get_width(self.line[: self.start_col])
            + "^" * _get_width(self.line[self.start_col : self.end_col])
        )


# AST
class Literal(NamedTuple):
    value: int
    metadata: Metadata


class FunRef(NamedTuple):
    rel: int
    metadata: Metadata


class ArgRef(NamedTuple):
    relA: AST
    relF: int
    metadata: Metadata


class FunDef(NamedTuple):
    body: AST
    metadata: Metadata


class FunCall(NamedTuple):
    fun: AST
    argv: tuple[AST, ...]
    metadata: Metadata


AST = Literal | FunRef | ArgRef | FunDef | FunCall


# Env
class Env(NamedTuple):
    funs: list[Closure]
    args: list[tuple[Value, ...]]


# Exception
class UnsuspectedHangeulError(Exception):
    def __init__(
        self,
        err: ErrorValue,
    ):
        message = ""
        for metadata in err.metadatas:
            message += (
                f"{metadata.filename} {metadata.line_no+1}번줄 "
                f"{metadata.start_col + 1}~{metadata.end_col+1}번째 글자:\n"
                f"{str(metadata)}\n"
            )
        super().__init__(f"{message}\n{err.message}")
        self.err = err  # Exposed to user code


# Values
@dataclasses.dataclass
class Integer:
    value: int

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Number) and self.value == other.value

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(self.value)
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return str(self.value)
        yield


@dataclasses.dataclass
class Float:
    value: float

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Number) and self.value == other.value

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(self.value)
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return str(self.value)
        yield


@dataclasses.dataclass
class Boolean:
    value: bool

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(("평범한 한글/참거짓", self.value))
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return str(self.value)
        yield


@dataclasses.dataclass
class List:
    value: tuple[Value, ...]
    _key: int | None = dataclasses.field(
        default=None, init=False, compare=False
    )
    _format: str | None = dataclasses.field(
        default=None, init=False, compare=False
    )

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        if self._key is None:
            keys: list[int] = []
            for item in self.value:
                item = yield item
                keys.append((yield from item.as_key()))
            self._key = hash(("평범한 한글/목록", tuple(keys)))
        return self._key

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        if self._format is None:
            formatted: list[str] = []
            for item in self.value:
                item = yield item
                formatted.append((yield from item.format()))
            self._format = "[" + ", ".join(formatted) + "]"
        return self._format


@dataclasses.dataclass
class String:
    value: str

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(("평범한 한글/문자열", self.value))
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return f"'{self.value}'"
        yield


@dataclasses.dataclass
class Bytes:
    value: bytes
    _format: str | None = dataclasses.field(
        default=None, init=False, compare=False
    )

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(("평범한 한글/바이트열", self.value))
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        if self._format is None:
            arg = "".join(rf"\x{b:02X}" for b in self.value)
            self._format = f"b'{arg}'"
        return self._format
        yield


@dataclasses.dataclass
class ErrorValue:
    metadatas: tuple[Metadata, ...]
    message: str
    value: tuple[StrictValue, ...]
    _key: int | None = dataclasses.field(
        default=None, init=False, compare=False
    )
    _format: str | None = dataclasses.field(
        default=None, init=False, compare=False
    )

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        if self._key is None:
            keys: list[int] = []
            for item in self.value:
                item = yield item
                keys.append((yield from item.as_key()))
            self._key = hash(("평범한 한글/예외", tuple(keys)))
        return self._key

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        if self._format is None:
            formatted: list[str] = []
            for item in self.value:
                item = yield item
                formatted.append((yield from item.format()))
            self._format = "<예외: [" + ", ".join(formatted) + "]>"
        return self._format


@dataclasses.dataclass
class Nil:
    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(("평범한 한글/빈값", None))
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return "Nil"
        yield


def _to_int_if_possible(num: float) -> int | float:
    if math.isfinite(num) and math.isclose(num, int(num), abs_tol=1e-16):
        return int(num)
    return num


@dataclasses.dataclass
class Complex:
    value: complex

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Number) and self.value == other.value

    def __str__(self):
        re = _to_int_if_possible(self.value.real)
        im = _to_int_if_possible(self.value.imag)
        re_str = str(re) + ("" if im < 0 else "+")
        minus_str = "-" if im < 0 else ""
        im_str = "" if abs(im) == 1 else str(abs(im))
        return "{}{}{}i".format(re_str if re else "", minus_str, im_str)

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(self.value)
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return str(self)
        yield


class Dict:
    def __init__(
        self,
        table: typing.Sequence[tuple[StrictValue, int, Value]],
    ):
        self.table = tuple(table)
        self.mapping = {k: v for _, k, v in table}
        self._key = None
        self._format = None

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        if self._key is None:
            keys: list[tuple[int, int]] = []
            for key, value in self.mapping.items():
                v = yield from (yield value).as_key()
                keys.append((key, v))
            self._key = hash(("평범한 한글/사전", frozenset(keys)))
        return self._key

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        if self._format is None:
            formatted: list[tuple[str, str]] = []
            for orig_key, key, value in self.table:
                del key  # Unused
                k = yield from (yield orig_key).format()
                v = yield from (yield value).format()
                formatted.append((k, v))
            formatted.sort(key=lambda pair: pair[0])
            self._format = (
                "{" + ", ".join(f"{k}: {v}" for k, v in formatted) + "}"
            )
        return self._format


class IO:
    def __init__(
        self,
        inst: str,
        argv: typing.Sequence[Value],
        continuation: typing.Callable[
            [
                typing.Callable[
                    [IO],
                    typing.Generator[Value, StrictValue, NonIOStrictValue],
                ]
            ],
            EvalContext,
        ],
    ):
        self._inst = inst
        self._argv = argv
        self.continuation = continuation
        self._key = None

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        if self._key is None:
            keys: list[int] = []
            for item in self._argv:
                item = yield item
                keys.append((yield from item.as_key()))
            self._key = hash(("평범한 한글/드나듦", self._inst, tuple(keys)))
        return self._key

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return f"<드나듦 {self._inst}>"
        yield


@dataclasses.dataclass
class BuiltinFunction:  # Is not a value
    literal: Literal


class Function(abc.ABC):
    def __init__(self, adj: str = ""):
        self._str = f"<{adj} 함수>"

    def as_key(self) -> typing.Generator[Value, StrictValue, int]:
        return hash(("평범한 한글/함수", id(self)))
        yield

    def format(self) -> typing.Generator[Value, StrictValue, str]:
        return self._str
        yield

    @abc.abstractmethod
    def __call__(
        self, metadata: Metadata, argv: typing.Sequence[Value]
    ) -> EvalContext:
        raise NotImplementedError


class Closure(Function):
    def __init__(self, body: AST, env: Env):
        self.body = body
        self.env = env
        super().__init__(f"깊이 {len(env.args)}에서 생성된")

    def __call__(
        self, metadata: Metadata, argv: typing.Sequence[Value]
    ) -> EvalContext:
        del metadata  # Unused
        canned_funs, canned_args = self.env
        new_env = Env(canned_funs, canned_args + [tuple(argv)])
        return Expr(self.body, new_env)
        yield


# Intermediate value
class CacheBox:
    def __init__(self):
        self._value: StrictValue | UnsuspectedHangeulError | None = None
        self.requestor: CacheBox | None = None

    @property
    def value(self):
        return self._value

    def resolve(self, value: StrictValue | UnsuspectedHangeulError):
        self._value = value
        if self.requestor:
            self.requestor.resolve(value)


class Expr:
    """Represents not yet evaluated value."""

    def __init__(self, expr: AST, env: Env):
        self.expr = expr
        self.env = env
        self.cache_box = CacheBox()


# Collection
Real = Integer | Float
Number = Real | Complex
Sequence = List | String | Bytes
Callable = Function | Boolean | Sequence | Dict | Complex | ErrorValue

NonIOStrictValue = Number | Callable | Nil
StrictValue = NonIOStrictValue | IO
Value = StrictValue | Expr

EvalContext = typing.Generator[
    Value,
    StrictValue,
    Value,
]
Evaluation = typing.Callable[[Metadata, typing.Sequence[Value]], EvalContext]

File: pbhhg_js/abstractSyntax.ts
/** Abstract syntax and values. **/
import Complex from 'complex.js'
import stringWidth from 'string-width'
import { isclose } from './numbers.js'

export class Metadata {
  private str?: string

  constructor(
    public filename: string,
    public line_no: number,
    public start_col: number,
    public end_col: number,
    public line: string
  ) {}

  toString() {
    if (this.str == null) {
      this.str =
        this.line +
        '\n' +
        ' '.repeat(stringWidth(this.line.slice(0, this.start_col))) +
        '^'.repeat(stringWidth(this.line.slice(this.start_col, this.end_col)))
    }
    return this.str
  }
}

/* Parser */
export class Literal {
  constructor(public metadata: Metadata, public value: bigint) {}
}
export class FunRef {
  constructor(public metadata: Metadata, public rel: number) {}
}
export class ArgRef {
  constructor(
    public metadata: Metadata,
    public relA: AST,
    public relF: number
  ) {}
}
export class FunDef {
  constructor(public metadata: Metadata, public body: AST) {}
}
export class FunCall {
  constructor(public metadata: Metadata, public fun: AST, public argv: AST[]) {}
}
export type AST = Literal | FunRef | ArgRef | FunDef | FunCall

/* Interpreter */
export type IOUtils = {
  input(): string | undefined
  print(content: string): void
}
export type LoadUtils = {
  load(location: string): string
  isFile(location: string): boolean
  listdir(location: string): string[]
  joinPath(...parts: string[]): string
  normalizePath(path: string): string
}
export class Env {
  constructor(
    public funs: ClosureV[],
    public args: Value[][],
    public utils: LoadUtils
  ) {}
}

export class UnsuspectedHangeulError extends Error {
  constructor(public err: ErrorV) {
    let message = ''
    for (const metadata of err.metadatas) {
      message += `${metadata.filename} ${metadata.line_no + 1}번줄 `
      message += `${metadata.start_col + 1}~${metadata.end_col + 1}번째 글자:\n`
      message += metadata.toString() + '\n'
    }
    super(message)
  }
}

/* Values */
interface ValueBase {
  asKey(strict: StrictFn): string
}

function _isPossibleInt(num: number) {
  return isFinite(num) && isclose(num, Math.floor(num))
}

function _formatFloat(num: number) {
  if (isFinite(num)) return num.toString()
  if (isNaN(num)) return 'nan'
  return num > 0 ? 'inf' : '-inf'
}

export class IntegerV implements ValueBase {
  static typeName = '정수'
  constructor(public value: bigint) {}
  toString() {
    return this.value.toString()
  }
  asKey(strict: StrictFn) {
    return this.toString()
  }
}

export class FloatV implements ValueBase {
  static typeName = '실수'
  constructor(public value: number) {}
  toString() {
    const str = _formatFloat(this.value)
    const trailing = _isPossibleInt(this.value) ? '.0' : ''
    return str + trailing
  }
  asKey(strict: StrictFn) {
    return _formatFloat(this.value)
  }
}

export class ComplexV implements ValueBase {
  static typeName = '복소수'
  constructor(public value: Complex) {}

  toString() {
    const re = this.value.re
    const im = this.value.im
    const reStr = _formatFloat(re) + (im < 0 ? '' : '+')
    const imStr = im == 1 ? '' : im == -1 ? '-' : _formatFloat(im)
    return (re === 0 ? '' : reStr) + imStr + 'i'
  }

  asKey(strict: StrictFn): string {
    return this.value.im ? this.toString() : _formatFloat(this.value.re)
  }
}

export class BooleanV implements ValueBase {
  static typeName = '논릿값'
  constructor(public value: boolean) {}
  toString() {
    return this.value ? 'True' : 'False'
  }
  asKey(strict: StrictFn) {
    return this.toString()
  }
}

export class ListV implements ValueBase {
  static typeName = '목록'
  constructor(public value: Value[]) {}
  asKey(strict: StrictFn): string {
    const children = this.value.map((value) => strict(value).asKey(strict))
    return `[${children.join(', ')}]`
  }
}

export class StringV implements ValueBase {
  static typeName = '문자열'
  constructor(public value: string) {}
  toString() {
    return "'" + this.value + "'"
  }
  asKey(strict: StrictFn): string {
    return this.toString()
  }
}

export class BytesV implements ValueBase {
  static typeName = '바이트열'
  private str?: string
  constructor(public value: ArrayBuffer) {}

  formatByte(c: number) {
    const s = c.toString(16).toUpperCase()
    return '\\x' + ('0' + s).slice(-2)
  }

  toString() {
    if (!this.str) {
      const arr = Array.from(new Uint8Array(this.value))
      const formatted = arr.map(this.formatByte)
      this.str = "b'" + formatted.join('') + "'"
    }
    return this.str
  }

  asKey(strict: StrictFn): string {
    return this.toString()
  }
}

export class DictV implements ValueBase {
  static typeName = '사전'
  private _keys?: string[]
  private _values?: Value[]
  constructor(public value: Record<string, Value>) {}

  keys() {
    if (!this._keys) {
      this._keys = Object.keys(this.value).sort()
    }
    return this._keys
  }

  values() {
    if (!this._values) {
      this._values = this.keys().map((k) => this.value[k])
    }
    return this._values
  }

  asKey(strict: StrictFn): string {
    const values = this.values().map((v) => strict(v).asKey(strict))
    const pairs = this.keys().map((k, i) => k + ': ' + values[i])
    return '{' + pairs.join(', ') + '}'
  }
}

export class IOV implements ValueBase {
  static typeName = '드나듦'
  constructor(
    public inst: string,
    public argv: Value[],
    public continuation: (
      doIO: (ioValue: IOV) => NonIOStrictValue,
      ioUtils: IOUtils
    ) => StrictValue
  ) {}

  asKey(strict: StrictFn): string {
    const formatted = this.argv
      .map((value) => strict(value).asKey(strict))
      .join(', ')
    return `<드나듦 ${this.inst}: [${formatted}]>`
  }
}

export class NilV implements ValueBase {
  static typeName = '빈값'
  constructor() {}
  toString() {
    return 'Nil'
  }
  asKey(strict: StrictFn): string {
    return 'Nil'
  }
}

let FUNCTION_ID_GEN = 0
export abstract class FunctionV implements ValueBase {
  static typeName = '함수'
  private id: number
  protected str: string
  constructor(adj = '') {
    this.id = FUNCTION_ID_GEN++
    this.str = '<' + adj + 'Function>'
  }
  toString() {
    return this.str
  }
  asKey(strict: StrictFn): string {
    return this.str.replace('>', ' #' + this.id + '>')
  }
  abstract execute(metadata: Metadata, argv: Value[]): Value
}

export class ClosureV extends FunctionV {
  constructor(public body: AST, public env: Env) {
    super()
    this.str = '<Closure created at depth ' + this.env.args.length + '>'
  }

  execute(metadata: Metadata, argv: Value[]) {
    const newArgs = this.env.args.concat([argv])
    const newEnv = new Env(this.env.funs, newArgs, this.env.utils)
    return new ExprV(this.body, newEnv, null)
  }
}

export class BuiltinModuleV extends FunctionV {
  constructor(public module: Evaluation, name: string) {
    super()
    this.str = '<Builtin Module ' + name + '>'
  }

  execute(metadata: Metadata, argv: Value[]) {
    return this.module(metadata, argv)
  }
}

export class ErrorV implements ValueBase {
  static typeName = '예외'
  constructor(
    public metadatas: Metadata[],
    public message: string,
    public value: StrictValue[]
  ) {}
  asKey(strict: StrictFn) {
    const formatted = this.value.map(strict).join(', ')
    return `<예외: [${formatted}]>`
  }
}

export class ExprV {
  constructor(
    public expr: AST,
    public env: Env,
    public cache: StrictValue | UnsuspectedHangeulError | null
  ) {}
}

export const RealV = [IntegerV, FloatV] as const
export const NumberV = [...RealV, ComplexV] as const
export const SequenceV = [ListV, StringV, BytesV] as const
export const CallableV = [
  FunctionV,
  BooleanV,
  DictV,
  ComplexV,
  ErrorV,
  ...SequenceV,
] as const
export const AnyV = [...NumberV, ...CallableV, IOV, NilV] as const

export type StrictValueType = (typeof AnyV)[number]

export type NonIOStrictValue =
  | IntegerV
  | FloatV
  | ComplexV
  | ListV
  | StringV
  | BytesV
  | BooleanV
  | DictV
  | ErrorV
  | FunctionV
  | NilV
export type StrictValue = NonIOStrictValue | IOV
export type Value = StrictValue | ExprV

export type ProcFunctionalFn = (
  metadata: Metadata,
  fun: Value,
  generalCallable?: boolean
) => Evaluation
export type StrictFn = (value: Value) => StrictValue
export type Evaluation = (metadata: Metadata, args: Value[]) => Value
